### ISSUE TITLE
Preendering and prefetch-src

### ISSUE THREAD

**Issue by mfalken** [w3c/webappsec-csp#502]
https://jeremyroman.github.io/alternate-loading-modes/ is exploring specifying prerendering: where a document can trigger a prerender to another document, and a subsequent navigation can activate that prerendered page.

There has been some discussion on https://github.com/jeremyroman/alternate-loading-modes/issues/43 on how Prerendering should behave with respect to CSP.

In summary we think that `prefetch-src` should control whether a document can trigger a prerender to a particular URL.

But one consideration is what happens if the CSP of the page changes after prerender is triggered, for example:
- The `<script type=speculationrules>` appears before a `<meta>` element that disallows it.
- A `<meta>` element that disallows an existing prerender is added after the prerender started.
- The document uses`document.write()` which changes the document and its CSP completely.

The current CSP spec says "policies in meta elements are not applied to content which precedes them.". However I'm not sure what the implications of that model should be for a prerendered page. One interpretation is that the prerendered page was created before the CSP was present, so it should be allowed to continue as if CSP didn't change. Another interpretation is to cancel prerendering upon CSP change or to disallow activation if the CSP at activation-time disallows the prerender.

Any thoughts on what a good way to go is here? In general, is CSP changing over the document lifetime something we should try to support?

**Comment by antosart** [comment]
> One interpretation is that the prerendered page was created before the CSP was present, so it should be allowed to continue as if CSP didn't change.

I believe this is the only thing that makes sense in this case. At activation time we should probably do the 'navigate-to' check and disallow activation if that does not pass.

(As a parallel: if you do `window.open("about:blank")` and then dynamically add a CSP via a `<meta>` tag, the newly added CSP applies to the original document, but the popup inherits the CSP as snapshotted when the popup is created, so without the new CSP)

I don't think that `<meta>` tags have the goal of changing CSPs over the document lifetime. I rather believe that is a side-effect of allowing to deliver CSPs via meta tags.

**Comment by mfalken** [comment]
Thanks, that makes sense to me.

**Comment by camillelamy** [comment]
To be clear, I am concerned specifically about the case where the page uses document.write at one point in its lifetime to update its CSPs. In this case, the intent is clearly to modify the document, including its CSPs. If we didn't have document,.write, then yes I agree we should only check at prerender start time. However, the document.write case is less clear-cut to me.

**Comment by noamr** [comment]
I don't see how this is different between prerender and prefetch.
Since the response of the prerender/prefetch is not observable by the initiating document, the only protection here is against exfiltration - in which case, if the prerender/prefetch request already went out, I think it's OK.

**Comment by annevk** [comment]
I don't think we should be overly concerned about `document.write()`. It's not an API we generally endorse using anyway.
